resource/Resource.md

```diff
@@ -0,0 +1,154 @@
+# Resource API Overview
+The resource library primarily defines a type that captures information about the entity
+for which stats or traces are recorded. It further provides a framework for detection of
+resource information from the environment and progressive population as signals propagate
+from the core instrumentation library to a backend's exporter.
+
+## Resource type
+A `Resource` describes the entity for which a signal was collected through two fields:
+* `type`: a string which describes a well-known type of entity. It SHOULD be namespaced
```

Comment:
I'd only consider doing this in the form of standardized tag keys potentially – which can be done later on when we've a better grasp of how people are using resources in different environments.
Dedicated fields will inevitably make user-side configuration more complex and the overall system harder to understand.

Reply:
I'd imagine system will want to draw applicaton map with the names taken from resource identity. In this case having `name` or `role` will be helpful as UI shouldn't worry about knowing all the different "providers" of such information.

Alernative may be to define a well-known postfix for those properties so exporter can do generic translation of labels to strongly-defined fields .

Reply:
I see what you are getting at. I agree that in the system where the data ends up, one wants to have more structure to enable certain workflows.

This proposal does not extend as far as telling backend system to have a resource presentation in form of a type string and tag map. Instead, the specific exporters should convert the rather structure-less resource notion proposed here into something more structured.
For Stackdriver for example, the exporter is configured with a mapping that converts known resource type and tag keys to Stackdriver Monitored Resources (e.g. [k8s_container](https://cloud.google.com/monitoring/api/resources#tag_k8s_container)) which have fixed fields.
Exporters will by default do a mapping of tags set by well-known resource detectors.

Reply:
Makes sense. What I'm trying to accomplish is some hint from resource-auto-discovery logic to exporter on what is what. So perhaps some postfix in names that help determine whether this tag is about unique ID of instance or logical group of instances (role). Or have those two on the level with `type`. Otherwise every exporter will need to know about every resource auto-detector.

Reply:
I'd expect the number of active detectors in a given environment to be very low so that configuration isn't a particularly big problem. Plus, if most common environments are supported by default, only a small fraction of users has to do configuration at all.

Coming up with a canonical set of schemas (in the fashion of [schema.org](https://schema.org)) will eat lots of time and is hard to get right. If it only relieves a few lines if configuration for a minority of users, I'm not sure that's worth it.

Again Prometheus as a data point – the provided data has even weaker structure and no default behavior at all, i.e. all users have to write rather awkward configuration. Regardless, they largely consider this to be a feature.